src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -2749,6 +2878,160 @@ private void ensurePublishCapability(
         }
     }
 
+    private final class MqttConnectPayload
+    {
+        private int reasonCode;
+
+        private final MqttWillProperties willPropertiesRO = new MqttWillProperties();
+
+        private String16FW willTopic = null;
+        private OctetsFW willPayload = null;
+        private String16FW username = null;
+        private OctetsFW password = null;
+
+        private void reset()
+        {
+            this.reasonCode = NORMAL_DISCONNECT;
+            this.willPropertiesRO.reset();
+            this.willTopic = null;
+            this.willPayload = null;
+            this.username = null;
+            this.password = null;
+        }
+
+        private void decodePayload(
+            MqttConnectFW connect)
+        {
+            final int flags = connect.flags();
+            final OctetsFW payload = connect.payload();
+
+            final DirectBuffer buffer = payload.buffer();
+            final int offset = payload.offset();
+            final int limit = payload.limit();
+
+            int payloadOffset = offset;
+            if ((flags & WILL_FLAG_MASK) != 0)
+            {
+                willPropertiesRO.decodeProperties(buffer, payloadOffset, limit);
+                payloadOffset = mqttPropertiesRO.limit();
+
+                willTopic = willTopicRO.tryWrap(buffer, payloadOffset, limit);
+                payloadOffset = willTopicRO.limit();
+
+                willPayload = willPayloadRO.tryWrap(buffer, payloadOffset, limit);
+                payloadOffset = willPayloadRO.limit();
+            }
+
+            if ((flags & USERNAME_MASK) != 0)
+            {
+                username = usernameRO.tryWrap(buffer, payloadOffset, limit);
+                payloadOffset = usernameRO.limit();
+            }
+
+            if ((flags & PASSWORD_MASK) != 0)
+            {
+                password = passwordRO.tryWrap(buffer, payloadOffset, limit);
+                payloadOffset = passwordRO.limit();
+            }
+        }
+    }
+
+    private final class MqttWillProperties
+    {
+        private String topic;
+        private int reasonCode;
+
+        private int willDelayInterval = 0;
+        private MqttPayloadFormat payloadFormat = DEFAULT_FORMAT;
+        private int expiryInterval = DEFAULT_EXPIRY_INTERVAL;
+        private String16FW contentType = null;
+        private String16FW responseTopic = null;
+        private OctetsFW correlationData = null;
+
+        private void reset()
+        {
+            this.topic = null;
+            this.reasonCode = SUCCESS;
+            this.willDelayInterval = 0;
+            this.expiryInterval = DEFAULT_EXPIRY_INTERVAL;
+            this.contentType = null;
+            this.payloadFormat = DEFAULT_FORMAT;
+            this.responseTopic = null;
+            this.correlationData = null;
+        }
+
+        private void decodeProperties(
+            DirectBuffer payloadBuffer,
+            int payloadOffset,
+            int payloadLimit)
+        {
```

Comment:
Suggest combining these 2 methods into `decode(MqttConnectFW)`, and adding a return type of `MqttConnectPayload`.

Then the caller will do something like:
```
MqttWillProperties willProperties = mqttWillPropertiesRO.decode(buffer, offset, limit);
willProperties.topic ...
```

This would be a useful pattern to apply to other payload decoders as well.